firedrake/slate/slate.py

```diff
@@ -0,0 +1,619 @@
+                                      % (type(other), type(self)))
+        else:
+            other.__mul__(self)
+
+    def __neg__(self):
+        return Negative(self)
+
+    def __eq__(self, other):
+        """Determines whether two TensorBase objects are equal using their
+        associated keys.
+        """
+        return self._key() == other._key()
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def ufl_domain(self):
+        """This function returns a single domain of integration occuring
+        in the tensor.
+
+        The function will fail if multiple domains are found.
+        """
+        domains = self.ufl_domains()
+        assert all(domain == domains[0] for domain in domains), (
+            "All integrals must share the same domain of integration."
+        )
+        return domains[0]
+
+    def __str__(self):
+        """Returns a string representation."""
+        return self._output_string(self.prec)
+
+    @cached_property
+    def _hash_id(self):
+        """Returns a hash id for use in dictionary objects."""
+        return hash(self._key())
+
+    def __hash__(self):
+        return self._hash_id
+
+
+class Tensor(TensorBase):
+    """This class is a symbolic representation of a finite element tensor
+    derived from a bilinear or linear form. This class implements all
+    supported ranks of general tensor (rank-0, rank-1 and rank-2 tensor
+    objects). This class is the primary user-facing class that the SLATE
+    symbolic algebra supports.
+
+    :arg form: a :class:`ufl.Form` object.
+
+    A :class:`ufl.Form` is currently the only supported input of creating
+    a `slate.Tensor` object:
+
+        (1) If the form is a bilinear form, namely a form with two
+            :class:`ufl.Argument` objects, then the SLATE Tensor will be
+            a rank-2 Matrix.
+        (2) If the form has one `ufl.Argument` as in the case of a typical
+            linear form, then this will create a rank-1 Vector.
+        (3) A zero-form will create a rank-0 Scalar.
+
+    These are all under the same type `slate.Tensor`. The attribute `self.rank`
+    is used to determine what kind of tensor object is being handled.
+    """
+
+    prec = None
+
+    def __init__(self, form):
+        """Constructor for the Tensor class."""
+        if not isinstance(form, Form):
+            raise ValueError("Only UFL forms are acceptable inputs for creating terminal tensors.")
+
+        r = len(form.arguments())
+        if r not in (0, 1, 2):
+            raise NotImplementedError("Currently don't support tensors of rank %d." % r)
+
+        # Checks for positive restrictions on integrals
+        integrals = form.integrals()
+        mapper = CheckRestrictions()
+        for it in integrals:
+            map_integrand_dags(mapper, it)
+
+        super(Tensor, self).__init__()
+
+        self.form = form
+
+        # Generate integral domains
```

Comment:
These two bits of code just copy the implementation from `ufl.Form`.  So I think you can avoid this copy-paste by just implementing ufl_domains with `return self.form.ufl_domains()` and subdomain_data with `return self.form.subdomain_data()`.